### PR TITLE
List and Export linked files

### PR DESCRIPTION
**Pull request by leoheck**
This is an attempt to list unique linked files of the assembly.
I also export files into a zip package with the name of the file + `asm4_zip` in the assembly folder.

### CODE REVIEW

**Comment by hasecilu** [comment]
I tried to load the changes but no icons are loaded in the WB, there is something in `InitGui.py` that breaks it

**Comment by hasecilu** [comment]
Is anytree, I have it in my system but not in FreeCAD's enviroment

**Comment by leoheck** [comment]
Did you do/try this?
```bash
sudo -H pip install anytree
```

**Comment by leoheck** [comment]
I installed it this way and it was working.

**Comment by hasecilu** [comment]
Ok, I solved it copying the `anytree` folder to the A4 folder.

The icon of `List Linked Files` is white/light gray and very similiar the the BOM buttons, maybe change to a similar color as `List Linked Files (Tree)` to have them "grouped" visually? Both commands work fine. There are typos in the svg files they say: Liked.

I got some errors with `Export FIles` and I think it's because those files were used more than once in the Assembly, so the function tries to add them again.
```
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/T8_leadscrew/T8_leadscrew_500mm.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/SHF08/SHF08.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/FreeCAD-library/SHF08/SHF08.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/KFL08/KFL08.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/FreeCAD-library/KFL08/KFL08.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/Electronics.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/Board4electronics.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/step/Arduino_CNC_shield.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/step/Fuente_12V_10A.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/Slot_T_nuts/Slot_T_nuts.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/FreeCAD-library/Slot_T_nuts/Slot_T_nuts.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: cad/Spacer_20.FCStd
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/SW_X.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/step/Limit_SW_PCB.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/step/Limit_SW_PCB.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/Green.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FDM/Light.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/step/Green_indicator.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/NEMA/NEMA_17_with_connector.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/Steel L-Bracket for NEMA 17.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/Steel L-Bracket for NEMA 17.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: cad/FDM/NEMA17_support.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/Coupling/Beam-coupling-8mm-5mm.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/FreeCAD-library/Coupling/Beam-coupling-8mm-5mm.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/3D-printed_corner.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FDM/Big_Corner/Big_Corner.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/Slot_T_nuts/Slot_T_nuts.FCStd
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/Bed_full.FCStd
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/Bed_top.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/Plate_03.FCStd
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/Bed_bottom.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/SK08/SK08.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/KP08/KP08.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/8mmx500_rod.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/8mmx500_rod.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: cad/FreeCAD-library/T8_leadscrew/T8_leadscrew_500mm.FCStd
13:22:32  /snap/freecad-realthunder/70/usr/lib/python3.8/zipfile.py:1517: UserWarning: Duplicate name: 'cad/FreeCAD-library/T8_leadscrew/T8_leadscrew_500mm.FCStd'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
13:22:32  ASM4> [ZIP] Adding file: sub-assemblies/SW_Y.FCStd
13:22:32  ASM4> [ZIP] Adding file: cad/FDM/LSW_support_Y.FCStd
```

**Comment by hasecilu** [comment]
> Did you do/try this?
> 
> ```shell
> sudo -H pip install anytree
> ```

I have just tried this and did not work in both snap and AppImage

**Comment by leoheck** [comment]
hm, I thought those "UserWarning: Duplicate name:" were just a warning, this may not be the case. They have to be added to preserve the hierarchy when unpacking the zip. I could rename all items before inserting in the package, and then I did not have errors, but this is not the fix since they won't be renamed back to the original when unpacking the zip. I am investigating this thing.

**Comment by hasecilu** [comment]
They still manage to be packed in the zip files, even with the unzip command I get messages asking me to replace the same files.
Screenshot from 2022-10-06 14-26-55

**Comment by leoheck** [comment]
try to extract it and load the main assembly, please.

**Comment by hasecilu** [comment]
For some reason in the list the googley eyes, Spacer 6.35 and V slot profile are not detected and the file is not saved in the zip.

The assembly in the zip file is almost complete but due to above error there are errors

**Comment by leoheck** [comment]
Ah, the list of files I was using did not have unique files. It is fixed now.
Zip is being created. But there are some issues when opening the main assembly file.

**Comment by leoheck** [comment]
These files are not being restored

```
Linked file: ../cad/Spacer_6.35.FCStd
Linked file: ../cad/FreeCAD-library/2020_V-slot_Al_extrusion/2020x370_V_slot_profile.FCStd
Linked file: ../cad/Googly_eyes.FCStd 
```

But they exist
```bash
➜  diy-pcb-cnc git:(testing) ✗ find -name Googly_eyes.FCStd
./cad/Googly_eyes.FCStd

➜  diy-pcb-cnc git:(testing) ✗ find -name 2020x370_V_slot_profile.FCStd 
./cad/FreeCAD-library/2020_V-slot_Al_extrusion/2020x370_V_slot_profile.FCStd

➜  diy-pcb-cnc git:(testing) ✗ find -name Spacer_6.35.FCStd
./cad/Spacer_6.35.FCStd
```

Maybe my parser is not finding them.

**Comment by leoheck** [comment]
Yeah, they are not in the above list with 57 items.

**Comment by leoheck** [comment]
`Bed_full` has spacers
And these are links from links. 
I was not expecting nested links yet.
image

**Comment by hasecilu** [comment]
The `Import Part` from A4 imports the external files as `Links` and `Links` have the Element Count property so I changed it and that generates all those sub-links

**Comment by leoheck** [comment]
I found your Bed_top has issues.
What is the most updated code you have? Is it on master or is it still in testing branch?

image

**Comment by hasecilu** [comment]
> These files are not being restored
> 
> ```
> Linked file: ../cad/Spacer_6.35.FCStd
> Linked file: ../cad/FreeCAD-library/2020_V-slot_Al_extrusion/2020x370_V_slot_profile.FCStd
> Linked file: ../cad/Googly_eyes.FCStd 
> ```

Those files had the Element Count property used.

**Comment by leoheck** [comment]
but look what my FC says
image

**Comment by leoheck** [comment]
There is no linked object
image

**Comment by leoheck** [comment]
ah, the files is missing, please, check if it is committed.

**Comment by hasecilu** [comment]
It is in the repo but it's not in the zip file

**Comment by leoheck** [comment]
@hasecilu took some time but now it exports all files of the main assembly.
Now, considering your repo, if you try to export a sub-assembly. It does not work yet. Some paths needs to be added, since the current flow only handles nested files/folders and not files and folders that are not inside of the same folder...

Example of that would be Bed_top.FCStd where most of the files comes from outside `../cad`

**Comment by hasecilu** [comment]
Cool, it's working!

> An example of that would be Bed_top.FCStd where most of the files come from outside ../cad

I'm not sure if there are file location guidelines for (sub) assemblies, moving `cad` inside of `sub-assemblies` could be a workaround but don´t know if there is a way to move linked files without breaking the links in assemblies.

**Comment by leoheck** [comment]
I was thinking that I could export the parent folder that has both assembly and its parts.

So if the assembly and related files are like this:
- `myass.FCSd`
- `../partslib/part1.FCstd`

Then it would normalize paths, considering the folder that has both (all) the parts.
It may work even if the paths are not related, I guess.

Also, some things to do
- Improve toolbar images
- Add a dialog so the user can choose the zip name and path while exporting
- Allow this to work with any kind of assembly (assemblies that do not have parts nested).

**Comment by leoheck** [comment]
Hey, @hasecilu now it exports assemblies with files in other folders. So it works with the Bed_top.FCSd the way I was thinking before. And the elegant thing would be to add a symlink to the exported assembly inside the root folder when the main assembly is not in the root folder. I will implement this tomorrow.

**Comment by hasecilu** [comment]
Tested and working!
> And the elegant thing would be to add a symlink to the exported assembly inside the root folder when the main assembly is not in the root folder. I will implement this tomorrow.

Sounds nice, could be useful when a assembly is made of many subassemblies and you need to find the main one.

**Comment by leoheck** [comment]
I got it working, but it happens that FreeCad can not handle symlinks since the internal links won't be adjusted if a symlink is being used. It could be a bug on Freecad or just something that was not thought of yet.

image

Then since symlink is in another place, the linked files cannot be found anymore.
image

Freecad could identify a symlink is being used to load the file from the original place, instead.

**Comment by leoheck** [comment]
I've pushed the code since it is harmless to the export process. 
If the assembly being exported is not in the root path of the zip package, it creates a symlink (you can test it with the `Bed_top`)
The only issue is that Freecad opens it but the linked files are not in the right place anymore. (this should be fixed on Freecad side, I gues)

**Comment by hasecilu** [comment]
In `Link`>`Linked object property` the path of the file is absolute and depending where you extract the zip file the absolute path of the exported zip file will change.

In my case I extracted in `.../DIY-PCB-CNC/sub-assemblies/Bed_top_asm4/` but the symlink is expecting files in another directory `.../DIY-PCB-CNC/sub-assemblies/cad/Plate_03.FCStd`.

If I extract in: `$HOME/Documents/test/Bed_top_asm4` it's expecting files in  `$HOME/Documents/test/cad/Plate_03.FCStd`.

Basically is skipping the `Bed_top_asm4` folder which is where the zip file was extracted.

I think the new icons need a color bg, the green and yellow of before were fine.
Screenshot from 2022-10-10 11-41-45

**Comment by leoheck** [comment]
Ah, I will have to tweak paths a bit. This feedback is what I was expecting, thanks. 

The icons are a bit blurry still, I may have to improve them better, just have to learn how. But check this new set for now, please. 
image

**Comment by hasecilu** [comment]
Probably they are blurry because of the icon size, when bigger they look crisper
Screenshot from 2022-10-10 12-35-45

**Comment by leoheck** [comment]
> In my case I extracted in .../DIY-PCB-CNC/sub-assemblies/Bed_top_asm4/ but the symlink is expecting files in another directory .../DIY-PCB-CNC/sub-assemblies/cad/Plate_03.FCStd.

Ah, yes. This is the issue I am talking about.
When you open a symlink, Freecad should follow the link, and then open the original file since it may have Linked files that have hardcoded or fixed paths.

**Comment by leoheck** [comment]
So the symlink is being created correctly. However, Freecad cannot open it fine, when it has Linked parts. Unless you are talking about a different problem.

**Comment by hasecilu** [comment]
> So the symlink is being created correctly. However, Freecad cannot open it fine, when it has Linked parts. Unless you are talking about a different problem.

It's the same issue. It looks like there is no differentiation between the files:
```
$ file ass.FCStd
ass.FCStd: symbolic link to subass/ass.FCStd
$ file subass/ass.FCStd
subass/ass.FCStd: Zip archive data, at least v2.0 to extract, compression method=deflate
```
